File: mirea/cources/2026-2027/aii-august/05/contest/C1.cpp
#include <bits/stdc++.h>
using namespace std;
// операции по простому модулю:
const int mod = 998244353;
// умножение:
int mulmod(int a, int b) { 
    return int(a * 1LL * b % mod);
}
// возведение в степень:
int binpow(int a, int64_t n) {
    int r = 1;
    while (n > 0) {
        if (n % 2 == 1)
            r = int(r * 1LL * a % mod);
        a = int(a * 1LL * a % mod);
        n >>= 1;
    }
    return r;
}
// обратный:
int invmod(int b) {
    return binpow(b, mod-2);
}
// деление:
int divmod(int a, int b) {
    return mulmod(a, invmod(b));
}
// решение задачи:
void solve() {
    int q; cin >> q;
    stack<int> st; // стек запросов в порядке обработки
    map<int, int> cnt; // количество каждого простого числа
    int answ = 1; // текущее количество делителей
    while (q --> 0) {
        char t; cin >> t;
        if (t == '+') {
            // умножение на новое число: увеличиваем количество и обновляем ответ
            int p; cin >> p;
            st.push(p);
            int prev = cnt[p]++;
            answ = divmod(answ, prev+1); // делим на то, что было
            answ = mulmod(answ, prev+2); // умножаем на то, что стало
        } else {
            // отмена операции: уменьшаем количество и обновляем ответ
            int p = st.top();
            st.pop();
            int prev = cnt[p]--;
            answ = divmod(answ, prev+1); // делим на то, что было
            answ = mulmod(answ, prev);   // умножаем на то, что стало
        }
        // выводим ответ:
        cout << answ << ' ';
    }
    cout << '\n';
}
main() {
    ios::sync_with_stdio(false);
    cin.tie(0);
    int tt; cin >> tt;
    while (tt --> 0) solve();
}

File: mirea/cources/2026-2027/aii-august/05/contest/C1.py
# Время: 765 мс, Память: 16660 КБ
# операции по простому модулю:
mod = 998244353
# умножение:
def mulmod(a, b):
    return a * b % mod
# обратный:
def invmod(b):
    return pow(b, -1, mod)
# деление:
def divmod(a, b):
    return mulmod(a, invmod(b))
# решение задачи:
def solve():
    q = int(input())
    st = [] # стек запросов в порядке обработки
    cnt = {} # количество каждого простого числа
    answ = 1 # текущее количество делителей
    while q > 0:
        s = input().split()
        t = s[0]
        if t == '+':
            # умножение на новое число: увеличиваем количество и обновляем ответ
            p = int(s[1])
            st.append(p)
            prev = cnt.get(p, 0)
            cnt[p] = prev + 1
            answ = divmod(answ, prev + 1) # делим на то, что было
            answ = mulmod(answ, prev + 2) # умножаем на то, что стало
        else:
            # отмена операции: уменьшаем количество и обновляем ответ
            p = st[-1]
            st.pop()
            prev = cnt[p]
            cnt[p] = prev - 1
            answ = divmod(answ, prev + 1) # делим на то, что было
            answ = mulmod(answ, prev) # умножаем на то, что стало
        # выводим ответ:
        print(answ, end=" ")
        q -= 1
    print()
tt = int(input())
while tt > 0:
    solve()
    tt -= 1